pkg/security/probe/load_controller.go

```diff
@@ -0,0 +1,203 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-2020 Datadog, Inc.
+
+// +build linux_bpf
+
+package probe
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/DataDog/datadog-go/statsd"
+	"github.com/hashicorp/golang-lru/simplelru"
+
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+// ByCount is used to order a list of EventCounter by count
+type ByCount []*EventCounter
+
+// Len returns the length of the list of EventCounter
+func (c ByCount) Len() int {
+	return len(c)
+}
+
+// Less returns true if the EventCounter at index i has collected more events than the EventCounter at index j
+func (c ByCount) Less(i, j int) bool {
+	return c[i].GetCount() < c[j].GetCount()
+}
+
+// Swap swaps the EventCounter at index i and j
+func (c ByCount) Swap(i, j int) {
+	c[i], c[j] = c[j], c[i]
+}
+
+// EventCounter is used to count the number of events retrieved from kernel space for a given pid and event type
+type EventCounter struct {
+	Pid           uint32
+	Event         EventType
+	OrderingCount uint64
+	Count         *uint64
+}
+
+// GetCount returns the current count of events
+func (ec *EventCounter) GetCount() uint64 {
+	if ec.OrderingCount == 0 {
+		ec.OrderingCount = atomic.SwapUint64(ec.Count, 0)
+	}
+	return ec.OrderingCount
+}
+
+// ResetOrderingCount resets the cached count used to order a list of EventCounter
+func (ec *EventCounter) ResetOrderingCount() {
+	ec.OrderingCount = 0
+}
+
+type eventCounterLRUKey struct {
+	Pid   uint32
+	Event EventType
+}
+
+// LoadController is used to monitor and control the pressure put on the host
+type LoadController struct {
+	sync.RWMutex
+	probe           *Probe
+	counters        *simplelru.LRU
+	orderedCounters []*EventCounter
+	statsdClient    *statsd.Client
+
+	EventsCountThreshold uint64
+	DiscarderTimeout     time.Duration
+	ControllerPeriod     time.Duration
+}
+
+// NewLoadController instantiates a new load controller
+func NewLoadController(probe *Probe, statsdClient *statsd.Client) (*LoadController, error) {
+	lru, err := simplelru.NewLRU(10000, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	lc := &LoadController{
+		probe:                probe,
+		counters:             lru,
+		statsdClient:         statsdClient,
+		EventsCountThreshold: probe.config.LoadControllerEventsCountThreshold,
+		DiscarderTimeout:     probe.config.LoadControllerDiscarderTimeout,
+		ControllerPeriod:     probe.config.LoadControllerControlPeriod,
+	}
+	return lc, nil
+}
+
+// Count increments the event counter of the provided event type and pid
+func (lc *LoadController) Count(eventType EventType, pid uint32) {
+	if eventType >= maxEventType {
```

Comment:
nit: this can probably be moved as we would not land here with an invalid eventType